### PR TITLE
Refactor MIN/MAX aggregate functions for istore and bigistore

### PR DESCRIPTION
**Pull request by chochkov**
Introduces custom transition state to MAX/MIN the same as on the SUM aggregate.

@rapimo as discussed this implements the base branch approach to MAX/MIN. I managed to reproduce the performance increase with an average `180ms` for version `0.1.1` over ~ 260K istores v.s. average of `102ms` for version `0.1.2` over the same set.

### CODE REVIEW

File: istore--0.1.1--0.1.2.sql [hunk 1]
```diff
@@ -1,38 +1,101 @@
...
  IMMUTABLE
 AS 'istore', $function$istore_sum_transfn$function$;
+
 ----aggregates----
 DROP AGGREGATE IF EXISTS sum (bigistore);
 CREATE AGGREGATE sum(bigistore) (
   SFUNC = public.istore_sum_transfn,
   STYPE = internal,
-  FINALFUNC = istore_sum_finalfn
+  FINALFUNC = istore_agg_finalfn
```
Comment (rapimo) [inline istore--0.1.1--0.1.2.sql]: this is probably not what we want

File: istore--0.1.1--0.1.2.sql [hunk 2]
```diff
@@ -1,38 +1,101 @@
...
 );
 
-
 ----
 DROP AGGREGATE IF EXISTS sum (istore);
 CREATE AGGREGATE sum(istore) (
   SFUNC = public.istore_sum_transfn,
   STYPE = internal,
-  FINALFUNC = istore_sum_finalfn
+  FINALFUNC = istore_agg_finalfn
```
Comment (rapimo) [inline istore--0.1.1--0.1.2.sql]: same here

File: istore--0.1.2.sql [hunk 3]
```diff
@@ -210,17 +220,21 @@ CREATE AGGREGATE SUM (
...
 );
 
-CREATE AGGREGATE MIN(istore) (
-    sfunc = istore_val_smaller,
-    stype = istore
+CREATE AGGREGATE MIN (
+    sfunc = istore_min_transfn,
+    basetype = istore,
+    stype = internal,
+    finalfunc = istore_agg_finalfn
```
Comment (rapimo) [inline istore--0.1.2.sql]: so MIN/MAX(istore) returns bigistore ? Is that what we want?

**Comment by rapimo** [comment]
this adds a lot code duplication
I'd suggest adding `istore_agg_internal(ISAggState *state, Datum,AggType)`
Where AggType is be
```
typedef enum
{
	SUM,
	MIN,
	MAX
} AggType;
```

also only sum_finalfn returns `bigistore` for both `istore` and `bigistore`
min and max should keep the type as overflow can not happen here.

That includes that we need separate state for istore and bigistore as the pairs type
differ here. An alternative would be to convert on finalfn copying pair by pair then 
instead of memcpy but a loop over pairs is needed anyway to get the buflen so these
steps can easily be combined.

**Comment by chochkov** [comment]
@rapimo thanks, i'll implement these comments.

**Comment by chochkov** [comment]
@rapimo @zeitgeist so this 

> An alternative would be to convert on finalfn copying pair by pair then 
instead of memcpy but a loop over pairs is needed anyway to get the buflen so these
steps can easily be combined.

is implemented here. The only duplication now is in the transformation functions - we have 6 - one per aggregation type and istore type. However they only handle arguments and check context, the `internal_agg` functions are only two - one per istore type. 

If there're improvement suggestions, please let me know.

File: src/istore_agg.c [hunk 4]
```diff
@@ -240,28 +232,157 @@ bigistore_sum_transfn(PG_FUNCTION_ARGS)
...
+
+    if (!AggCheckCallContext(fcinfo, &agg_context))
+        elog(ERROR, "aggregate function called in non-aggregate context");
+
+    if (PG_ARGISNULL(1) && PG_ARGISNULL(0)) PG_RETURN_NULL();
+
+    state = PG_ARGISNULL(0) ? state_init(agg_context) : (ISAggState *) PG_GETARG_POINTER(0);
+
+    if (PG_ARGISNULL(1)) PG_RETURN_POINTER(state);
+
```
Comment (rapimo) [inline src/istore_agg.c]: this duplication could be avoided with the help of a macro

File: istore--0.1.1--0.1.2.sql [hunk 5]
```diff
@@ -1,38 +1,107 @@
...
+ RETURNS internal
+ LANGUAGE c
+ IMMUTABLE
+AS 'istore', $function$bigistore_max_transfn$function$;
+
+----
+CREATE FUNCTION istore_min_transfn(internal, bigistore)
+ RETURNS internal
+ LANGUAGE c
+ IMMUTABLE
```
Comment (zeitgeist) [inline istore--0.1.1--0.1.2.sql]: btw shouldn't the aggregate functions be declared `STRICT`. This would end the `PG_ARGISNULL(1)` calls below.
Speaking of which, there are other places with `PG_ARGISNULL` calls that appear redundant to me (i.e. `istore_array_add` to name one such function – this is not part of the PR though, don't get me wrong!).

Comment (zeitgeist) [inline istore--0.1.1--0.1.2.sql]: to put this into reference a bit:
> If the state transition function is declared "strict", then it cannot be called with null inputs. With such a transition function, aggregate execution behaves as follows. Rows with any null input values are ignored (the function is not called and the previous state value is retained). If the initial state value is null, then at the first row with all-nonnull input values, the first argument value replaces the state value, and the transition function is invoked at each subsequent row with all-nonnull input values. This is handy for implementing aggregates like max. Note that this behavior is only available when state_data_type is the same as the first arg_data_type. When these types are different, you must supply a nonnull initial condition or use a nonstrict transition function.
(from https://www.postgresql.org/docs/9.4/static/sql-createaggregate.html)

Comment (zeitgeist) [inline istore--0.1.1--0.1.2.sql]: OK, so reading into the PR a bit more, I now get that I'm completely wrong here. Please apologize the noise.

Comment (chochkov) [inline istore--0.1.1--0.1.2.sql]: @zeitgeist I had the same train of thought earlier, but the because the types are different we can't use the strict definition.

**Comment by chochkov** [comment]
@rapimo thanks for the idea - that's exactly what i needed and i did it.

File: src/istore_agg.c [hunk 6]
```diff
@@ -3,6 +3,19 @@
...
+                                                                                             \
+    if (!AggCheckCallContext(fcinfo, &agg_context))                                          \
+        elog(ERROR, "aggregate function called in non-aggregate context");                   \
+                                                                                             \
+    if (PG_ARGISNULL(1) && PG_ARGISNULL(0)) PG_RETURN_NULL();                                \
+                                                                                             \
+    state = PG_ARGISNULL(0) ? state_init(agg_context) : (ISAggState *) PG_GETARG_POINTER(0); \
+                                                                                             \
+    if (PG_ARGISNULL(1)) PG_RETURN_POINTER(state);                                           \
+
```
Comment (rapimo) [inline src/istore_agg.c]: did you try this out?
afaik this should not work unless you have some magic compiler our I'm misreading something here
see https://github.com/adjust/istore/blob/master/src/is_parser.h#L16
on how to do multiline macros

Comment (rapimo) [inline src/istore_agg.c]: ok forget what I just said I was thinking about something different

Comment (rapimo) [inline src/istore_agg.c]: it's probably best practice to add block though
```
#define AGG_TRANSFN_INIT \
{\
.... \
}
```
@roa or @zeitgeist can add their thoughts

Comment (chochkov) [inline src/istore_agg.c]: The problem is that if we put the macro within a block the `MemoryContext  agg_context;` declaration for example will be only available inside it. So we'll need to take it out of the macro.

I read some discussion on the topic of `{}` v.s. `do {} while(0)`, but I am also curious to hear what the others have to say.

Comment (chochkov) [inline src/istore_agg.c]: Or maybe we could pass a pointer to the relevant `*_agg_internal` function to the macro too - i'll experiment with it.

Comment (rapimo) [inline src/istore_agg.c]: I guess it's ok to leave it as is

Comment (chochkov) [inline src/istore_agg.c]: well.. if we ever put this in 

```
if (cond)
  AGG_TRANSF_INIT
else
   ..
```

it'll be broken..

Comment (chochkov) [inline src/istore_agg.c]: let me try and refactor it quickly

Comment (chochkov) [inline src/istore_agg.c]: i did here https://github.com/adjust/istore/pull/55/commits/72a6ee254d4b5d0150d9cac1d54de79595e9acdb

File: src/istore_agg.c [hunk 7]
```diff
@@ -3,18 +3,33 @@
...
+        switch (_istore)                                                                         \
+        {                                                                                        \
+            case BIG_ISTORE:                                                                     \
+                PG_RETURN_POINTER(bigistore_agg_internal(state, PG_GETARG_BIGIS(1), _agg));      \
+            case ISTORE:                                                                         \
+                PG_RETURN_POINTER(istore_agg_internal(state, PG_GETARG_IS(1), _agg));            \
+            default:                                                                             \
+                elog(ERROR, "unexpected istore type passed.");                                   \
+        }                                                                                        \
+    } while(0)
```
Comment (rapimo) [inline src/istore_agg.c]: @zeitgeist will the preprocessor handle the switch ?
@chochkov won't the compile complain about missing return ?
overall that looks a bit mehhh

Comment (zeitgeist) [inline src/istore_agg.c]: @rapimo since '_istore' is constant at macro invocation time, the compiler will just optimize the switch statement away

Comment (chochkov) [inline src/istore_agg.c]: we could try the inline function idea from @zeitgiest

Comment (rapimo) [inline src/istore_agg.c]: so `RETURN_AGG_TRANSFN(2, _agg)`
would compile to to a function without any return statement which the compiler would error right ?
which would be the same as having no default at all

**Comment by rapimo** [comment]
@chochkov @zeitgeist how about aa35471

File: src/istore_agg.c (hunk 7 above)
Comment (chochkov) [inline src/istore_agg.c]: @rapimo afaics `elog()` macro calls `abort()` eventually so that's why it works in this case - but i agree it's confusing

**Comment by chochkov** [comment]
it's 👍  by me @rapimo

File: src/istore_agg.c (hunk 7 above)
Comment (zeitgeist) [inline src/istore_agg.c]: i'm not 100% sure about this, but
1. the compiler should either error or not at all, since the case branches stay the same (syntactically speaking here) no matter what the input
2. the compiler probably doesn't because it realizes that the default branch is not taken
3. if there is no return statement in a function having a return value, the behavior is to return the value returned from the last function call. 

To catch this illegal assignment at compile time, one would have to enable it `-Werror=return-type` AFAICT.

The easiest way to go is to use a compile-time assertion like this:

```C
StaticAssertStmt(_istore == 0 || _istore == 1, "illegal istore type");
```
